enhancements/cvo/cluster-profiles.md

```diff
@@ -88,6 +88,9 @@ has been specified.
 Manifests may support inclusion in multiple profiles by including as many of these annotations
 as needed.
 
+Manifests with no explicit inclusion annotations implicitly belong to the `default` profile. 
+If no profile is configured, the cluster-version operator should use the `default` profile.
```

Comment:
Thinking this over some more, we may need a more involved reword.  Above in master we have:

> This would make the CVO render this manifest only when `CLUSTER_PROFILE=[identifier]`
has been specified.

But that's not actually true, because as the next master paragraph points out, you can list multiple profiles.  I suggest we reword this section to say something like:

> This environment variable would have to be specified in the CVO deployment.
> When no `CLUSTER_PROFILE=[identifier]` variable is specified, the `default` cluster profile
is in effect.  [NOTE: so I guess no need to echo this with a new line]
>
> The following annotation may be used to include manifests for a given profile:
>
> ```
> include.release.openshift.io/[identifier]=true
> ```
>
> Manifests may support inclusion in multiple profiles by including as many of these annotations
as needed.
> Manifests with any explicit inclusion annotations only belong to the declared profiles.
> Manifests with no explicit inclusion annotations implicitly belong to the `default` profile.
>
> The cluster-version operator will only reconcile manifests which belong to its configured cluster profile.

Also note that the implicit `default` here somewhat pushes back against @deads2k's [point about avoiding exclusion][1].  It means that you don't need to put in the awareness-raising work he wants for the default profile.  But it also means that we don't break compat with the existing default-only profile out of the gate.  I'm fine with that.  I'd also be ok with "too bad, you need to get everyone to annotate their existing manifests with `include.release.openshift.io/[identifier]=true` before teaching the CVO anything", if folks feel like the implicit `default` membership is too magical.

[1]: https://github.com/openshift/enhancements/pull/200#discussion_r375837903

Reply:
> It means that you don't need to put in the awareness-raising work he wants for the default profile. But it also means that we don't break compat with the existing default-only profile out of the gate.

I'd favour first getting the 'magic' default behaviour in CVO, then add 'default' awareness to manifests, and finally remove this magic behaviour when everyone has switched.

Reply:
I suggest we specify instead that "Manifests with no explicit inclusion annotations implicitly belong to all profiles." 

This would mean that all existing manifests, as well as all newly introduced manifests, are included in all profiles, unless someone specifically adds annotations to indicate otherwise. If an annotation is added to a manifest, then annotations will be needed for all profiles in which that manifest should be included.

This approach will help avoid accidental divergence between what functionality is included in different cluster profiles. Any new manifest will be included in all profiles, unless it has annotations for specific profiles. 

It will also mean that a new profile (such as a single-node-cluster profile for CRC, for which we are working on an enhancement PR) can be added and then modified incrementally, rather than requiring immediate PRs to be merged into every repo containing any manifests that should be included in that profile. 

My assumption here is that many or most manifests, at least initially, will not need to vary between the small set of profiles that exist, and thus will not require any annotations. But if that assumption proves invalid over time, once we've added explicit annotations to every manifest, we could then change the CVO to require an explicit annotation for any manifest to be included in any profile.

Reply:
> This would mean that all existing manifests, as well as all newly introduced manifests, are included in all profiles, unless someone specifically adds annotations to indicate otherwise... This approach will help avoid accidental divergence between what functionality is included in different cluster profiles

That diverges from [David's][1]:

> To the common counter argument of "but then it takes a lot of work to create a new profile and all the teams have to aware of it", I say "good!"...

If folks want to push back against all (but the default?) profile being explicitly opt-in for each manifest, I'm not going to stand in the way, but please go fight it out with @deads2k and report back here once you've reached a consensus ;).

[1]: https://github.com/openshift/enhancements/pull/200#discussion_r375837903

Reply:
Not looking for any fights - just for clarity on what we'll need to do to add and maintain the single-node profile for CRC, for which we are working on a separate enhancement PR. This profile will ideally remain identical to normal OCP deployments, except without HA, with a few features not important for application development left out, and with things tuned to minimize resource requirements.

Reply:
@deads2k 's previous comment seemed to call for explicitly talking new additions into joining each profile.  That's obviously more work, but his point is that thinking each manifest addition through is useful.  Including by default in the absence of explicit exclusion-action is also internally consistent.  CVO can easily implement either.  I want some kind of consensus, so we don't flop back and forth depending on who asked most recently.

Reply:
I still like my reasoning in https://github.com/openshift/enhancements/pull/200#discussion_r375837903.  Having to modify every manifest you want to include in a profile does a couple things
1. Requires a cost to adding a profile. This cost is updating a few hundred file in a few dozen repos.  While this seems large initially, the long term cost born by teams for each new topology is a much larger ongoing effort.  If it's not worth the few days to find and update, then the profile may not be valuable enough to support.
2. This exposes the profile to every component which can be impacted.  And because the repos are separate, each component owner will be explicitly aware and cannot claim ignorance in the future.  Component owners should question the change and ensure their component/operator can handle it.

To help provide an example and to be sure we can avoid accidental code defaults, I'm in support of actually updating all the existing manifests to include a reference to the default.  I'm even willing to help make those PRs and get them merged.

Reply:
I'm OK with @deads2k's reasoning on this, as long as getting patches merged to add annotations to manifests across all needed repositories does not turn out to be an insurmountable obstacle. We plan to have an enhancement PR for adding a single node profile for CRC in review soon. This will show that most manifests in OpenShift will need to be included in this profile, many with no profile-specific changes. I'd expect we'd need to merge a complete set of initial annotations early in development in order to allow the profile to be used to produce a cluster equivalent to the current CRC. Then additional PRs would be merged to further optimize the profile for CRC. Does this seem workable?

Reply:
@deads2k : #482 reminded me that this might still be in flight, and seems like [rhbz#1871890][1] still links a bunch of open PRs, mostly in the kube-API sphere of influence.  And seems like maybe there aren't even PRs up yet for many components outside the kube-API sphere of influence.  Any preliminary thoughts on whether this is something that can be accomplished with a single driving dev, vs. something where you need to call for teams to self-serve in aos-devel@, and then run around and poke folks who don't act on that call?

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1871890